designs/cloud_formation_api_gateway_support.md

```diff
@@ -0,0 +1,377 @@
+****Feature* #2*: Using a combination of AWS::ApiGateway::Resource with ApiGateway::Methods.**
+
+Although this approach is less common and more verbose, it is still used by some people while defining their resources. Tools 
+such as aws-cdk currently generate Resource and Methods ApiGateway CloudFormation Resources in their yaml instead of swagger. 
+
+```yaml
+UsersResource:
+  Type: AWS::ApiGateway::Resource
+  Properties:
+    RestApiId:!Ref RestApi
+    ParentId:
+      Fn::GetAtt:
+      - RestApi
+      - RootResourceId
+    PathPart: user
+UsersGet:
+  Type: AWS::ApiGateway::Method
+  Properties:
+    ResourceId: !Ref ApiGatewayResource
+    RestApiId: !Ref ApiGatewayRestApi
+    Integration:
+      Type: AWS
+      IntegrationHttpMethod: POST
+      Uri:
+        Fn::Join:
+        - ''
+        - - 'arn:aws:apigateway:'
+          - Ref: AWS::Region
+          - ":lambda:path/2015-03-31/functions/"
+          - Fn::GetAtt:
+            - Lambda
+            - Arn
+          - "/invocations"
+      IntegrationResponses: []
+    MethodResponses:
+    - ResponseModels:
+        application/json:
+          Ref: UsersModel
+      ResponseParameters:
+        method.response.header.Link: true
+      StatusCode: 200
+   ApiGatewayMethod:
+    Type: AWS::ApiGateway::Method
+    Properties:
+      HttpMethod: POST
+      Integration:
+        ConnectionType: INTERNET
+        IntegrationResponses:
+          - ResponseTemplates:
+              application/json: "{\"message\": \"OK\"}"
+            StatusCode: 200
+        RequestTemplates:
+          applicationjson: "{\"statusCode\": $input.json('$.statusCode'), \"message\": $input.json('$.message')}"
+
+```
+
+The SAM template requires people to use only a single RestApi to define all of their resources.
+
+### **AWS::ApiGateway::Stages**
+
+The stage allows for defining environments for different parts of the pipeline in CloudFormation. Since a single stage is attached to a deployment, the behavior of multiple deployment/stages locally is currently undefined, which is similar to the current support or syntax in the SAM template. 
+
+Currently, the SAM template supports only one stage/one deployment and does not /<Stage>/<Serivce Name> as customers could want. One approach is to support additional flags in environments and stage names so they can test the apis locally in multiple environments. This would create a dict such as {“dev”: [models], “prod”: [models]}. However, this may be unnecessary. 
+
+The AWS::ApiGateway::Stage will initially support the schema:
+
+```yaml
+Prod:
+Type: AWS::ApiGateway::Stage
+Properties:
+  StageName: Prod
+  Description: Prod Stage
+  Variables:
+    Stack: Prod
+  RestApiId: RestApi`
+```
+
+Since the stages can show up in any order as the code is iterating through the resources, There are two approaches to consider too quickly stitch the api and stages together. 
+
+*Approach #1:* 
+Arbitrarily pick one of the stages and use that on all the Api
```

Comment:
Picking arbitrarily is fine but we should make sure it is reproducible. Meaning if the customers gives the same template in two different command runs, the same stage should be selected, otherwise this could become frustrating to customers.

I think Arbitrarily here means we pick the first one in the template we find?